libsolidity/formal/SMTEncoder.cpp

```diff
@@ -755,10 +755,27 @@ void SMTEncoder::arrayIndexAssignment(Expression const& _expr, smt::Expression c
 		auto varDecl = identifierToVariable(*id);
 		solAssert(varDecl, "");
 
+		bool resetReferences = false;
 		if (varDecl->hasReferenceOrMappingType())
+		{
+			auto const& arrayType = dynamic_cast<ArrayType const*>(varDecl->type());
+			// Statically sized arrays are copied
```

Comment:
Is this so? Maybe I'm misunderstanding what case this is, but is this the following case, resp. does this work?
```
uint256[2] storage a;
uint256[2] storage b;
function f(uint256[2] storage x) internal {
  require(a[0] == 42);
  x[0] = 21;
  assert(a[0] == 42); // should fail since possibly x == a?
}
```

Reply:
Ah you're right, only memory ones are copied.

Reply:
But even they only for external calls, right? With internal calls, this doesn't copy either:
```
contract C {
    function f() public returns (bool) {
        uint8[1] memory y = [23];
        assert(y[0] == 23);
        g(y);
        assert(y[0] == 42);
        return true;
        
    }
    function g(uint8[1] memory x) internal {
        x[0] = 42;
    }
}
```

Reply:
And also this doesn't copy:
```
contract C {
    function f() public returns (bool) {
        uint8[1] memory a = [42];
        uint8[1] memory b = a;
        b[0] = 23;
        assert(a[0] == 23);
        return true;
    }
}
```
... it's really time to introduce ``copyof``- I keep rechecking this myself :-). Only memory->storage and vice-verse always has to copy.

Reply:
Yea I noticed that too when writing new tests but didn't have time to push. Will update tomorrow.

Reply:
Updated